### ISSUE TITLE
Improve ability to limit build cache upload size

### ISSUE THREAD

**Issue by DanielThomas** [gradle/gradle#13608]
We limit build cache uploads to 10MB, however the build cache implementation doesn't appear to deal well with the upload being short-circuited:

```
Could not store entry 2d499ed2c28b7f6f832b5cf09651add2 in remote build cache
org.gradle.caching.BuildCacheException: Broken pipe (Write failed)
        at org.gradle.caching.http.internal.HttpBuildCacheService.wrap(HttpBuildCacheService.java:190)
        at org.gradle.caching.http.internal.HttpBuildCacheService.store(HttpBuildCacheService.java:176)
        at org.gradle.caching.internal.controller.service.BaseBuildCacheServiceHandle.storeInner(BaseBuildCacheServiceHandle.java:93)
        at org.gradle.caching.internal.controller.service.OpFiringBuildCacheServiceHandle.access$001(OpFiringBuildCacheServiceHandle.java:35)
        at org.gradle.caching.internal.controller.service.OpFiringBuildCacheServiceHandle$2.run(OpFiringBuildCacheServiceHandle.java:71)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:395)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$RunnableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:387)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:242)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:150)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.run(DefaultBuildOperationExecutor.java:84)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.run(DelegatingBuildOperationExecutor.java:31)
        at org.gradle.caching.internal.controller.service.OpFiringBuildCacheServiceHandle.storeInner(OpFiringBuildCacheServiceHandle.java:68)
        at org.gradle.caching.internal.controller.service.BaseBuildCacheServiceHandle.store(BaseBuildCacheServiceHandle.java:86)
        at org.gradle.caching.internal.controller.DefaultBuildCacheController.lambda$store$1(DefaultBuildCacheController.java:182)
        at org.gradle.caching.local.internal.DefaultBuildCacheTempFileStore.withTempFile(DefaultBuildCacheTempFileStore.java:46)
        at org.gradle.caching.internal.controller.DefaultBuildCacheController.store(DefaultBuildCacheController.java:178)
        at org.gradle.internal.execution.steps.CacheStep.store(CacheStep.java:145)
        at org.gradle.internal.execution.steps.CacheStep.lambda$executeAndStoreInCache$6(CacheStep.java:137)
        at org.gradle.internal.Try$Success.ifSuccessfulOrElse(Try.java:187)
        at org.gradle.internal.execution.steps.CacheStep.executeAndStoreInCache(CacheStep.java:136)
        at org.gradle.internal.execution.steps.CacheStep.lambda$executeWithCache$2(CacheStep.java:112)
        at java.util.Optional.orElseGet(Optional.java:267)
        at org.gradle.internal.execution.steps.CacheStep.lambda$executeWithCache$3(CacheStep.java:112)
        at org.gradle.internal.Try$Success.map(Try.java:162)
        at org.gradle.internal.execution.steps.CacheStep.executeWithCache(CacheStep.java:81)
        at org.gradle.internal.execution.steps.CacheStep.execute(CacheStep.java:71)
        at org.gradle.internal.execution.steps.CacheStep.execute(CacheStep.java:43)
        at org.gradle.internal.execution.steps.StoreExecutionStateStep.execute(StoreExecutionStateStep.java:44)
        at org.gradle.internal.execution.steps.StoreExecutionStateStep.execute(StoreExecutionStateStep.java:33)
        at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:38)
        at org.gradle.internal.execution.steps.RecordOutputsStep.execute(RecordOutputsStep.java:24)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.executeBecause(SkipUpToDateStep.java:92)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.lambda$execute$0(SkipUpToDateStep.java:85)
        at java.util.Optional.map(Optional.java:215)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:55)
        at org.gradle.internal.execution.steps.SkipUpToDateStep.execute(SkipUpToDateStep.java:39)
        at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:76)
        at org.gradle.internal.execution.steps.ResolveChangesStep.execute(ResolveChangesStep.java:37)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:36)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsFinishedStep.execute(MarkSnapshottingInputsFinishedStep.java:26)
        at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:94)
        at org.gradle.internal.execution.steps.ResolveCachingStateStep.execute(ResolveCachingStateStep.java:49)
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:79)
        at org.gradle.internal.execution.steps.CaptureStateBeforeExecutionStep.execute(CaptureStateBeforeExecutionStep.java:53)
        at org.gradle.internal.execution.steps.ValidateStep.execute(ValidateStep.java:74)
        at org.gradle.internal.execution.steps.SkipEmptyWorkStep.lambda$execute$2(SkipEmptyWorkStep.java:78)
        at java.util.Optional.orElseGet(Optional.java:267)
        at org.gradle.internal.execution.steps.SkipEmptyWorkStep.execute(SkipEmptyWorkStep.java:78)
        at org.gradle.internal.execution.steps.SkipEmptyWorkStep.execute(SkipEmptyWorkStep.java:34)
        at org.gradle.internal.execution.steps.legacy.MarkSnapshottingInputsStartedStep.execute(MarkSnapshottingInputsStartedStep.java:39)
        at org.gradle.internal.execution.steps.LoadExecutionStateStep.execute(LoadExecutionStateStep.java:40)
        at org.gradle.internal.execution.steps.LoadExecutionStateStep.execute(LoadExecutionStateStep.java:28)
        at org.gradle.internal.execution.impl.DefaultWorkExecutor.execute(DefaultWorkExecutor.java:33)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.executeIfValid(ExecuteActionsTaskExecuter.java:194)
        at org.gradle.api.internal.tasks.execution.ExecuteActionsTaskExecuter.execute(ExecuteActionsTaskExecuter.java:186)
        at org.gradle.api.internal.tasks.execution.CleanupStaleOutputsExecuter.execute(CleanupStaleOutputsExecuter.java:114)
        at org.gradle.api.internal.tasks.execution.FinalizePropertiesTaskExecuter.execute(FinalizePropertiesTaskExecuter.java:46)
        at org.gradle.api.internal.tasks.execution.ResolveTaskExecutionModeExecuter.execute(ResolveTaskExecutionModeExecuter.java:62)
        at org.gradle.api.internal.tasks.execution.SkipTaskWithNoActionsExecuter.execute(SkipTaskWithNoActionsExecuter.java:57)
        at org.gradle.api.internal.tasks.execution.SkipOnlyIfTaskExecuter.execute(SkipOnlyIfTaskExecuter.java:56)
        at org.gradle.api.internal.tasks.execution.CatchExceptionTaskExecuter.execute(CatchExceptionTaskExecuter.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.executeTask(EventFiringTaskExecuter.java:77)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:55)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter$1.call(EventFiringTaskExecuter.java:52)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:409)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$CallableBuildOperationWorker.execute(DefaultBuildOperationExecutor.java:399)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor$1.execute(DefaultBuildOperationExecutor.java:157)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:242)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.execute(DefaultBuildOperationExecutor.java:150)
        at org.gradle.internal.operations.DefaultBuildOperationExecutor.call(DefaultBuildOperationExecutor.java:94)
        at org.gradle.internal.operations.DelegatingBuildOperationExecutor.call(DelegatingBuildOperationExecutor.java:36)
        at org.gradle.api.internal.tasks.execution.EventFiringTaskExecuter.execute(EventFiringTaskExecuter.java:52)
        at org.gradle.execution.plan.LocalTaskNodeExecutor.execute(LocalTaskNodeExecutor.java:41)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:356)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$InvokeNodeExecutorsAction.execute(DefaultTaskExecutionGraph.java:343)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:336)
        at org.gradle.execution.taskgraph.DefaultTaskExecutionGraph$BuildOperationAwareExecutionAction.execute(DefaultTaskExecutionGraph.java:322)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.lambda$run$0(DefaultPlanExecutor.java:127)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.execute(DefaultPlanExecutor.java:191)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.executeNextNode(DefaultPlanExecutor.java:182)
        at org.gradle.execution.plan.DefaultPlanExecutor$ExecutorWorker.run(DefaultPlanExecutor.java:124)
        at org.gradle.internal.concurrent.ExecutorPolicy$CatchAndRecordFailures.onExecute(ExecutorPolicy.java:64)
        at org.gradle.internal.concurrent.ManagedExecutorImpl$1.run(ManagedExecutorImpl.java:48)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
        at org.gradle.internal.concurrent.ThreadFactoryImpl$ManagedThreadRunnable.run(ThreadFactoryImpl.java:56)
        at java.lang.Thread.run(Thread.java:748)
Caused by: java.net.SocketException: Broken pipe (Write failed)
        at java.net.SocketOutputStream.socketWrite0(Native Method)
        at java.net.SocketOutputStream.socketWrite(SocketOutputStream.java:111)
        at java.net.SocketOutputStream.write(SocketOutputStream.java:155)
        at sun.security.ssl.OutputRecord.writeBuffer(OutputRecord.java:431)
        at sun.security.ssl.OutputRecord.write(OutputRecord.java:417)
        at sun.security.ssl.SSLSocketImpl.writeRecordInternal(SSLSocketImpl.java:894)
        at sun.security.ssl.SSLSocketImpl.writeRecord(SSLSocketImpl.java:865)
        at sun.security.ssl.AppOutputStream.write(AppOutputStream.java:123)
        at org.apache.http.impl.io.SessionOutputBufferImpl.streamWrite(SessionOutputBufferImpl.java:124)
        at org.apache.http.impl.io.SessionOutputBufferImpl.flushBuffer(SessionOutputBufferImpl.java:136)
        at org.apache.http.impl.io.SessionOutputBufferImpl.write(SessionOutputBufferImpl.java:167)
        at org.apache.http.impl.io.ContentLengthOutputStream.write(ContentLengthOutputStream.java:113)
        at com.google.common.io.ByteStreams.copy(ByteStreams.java:112)
        at com.google.common.io.ByteSource.copyTo(ByteSource.java:246)
        at org.gradle.caching.internal.controller.service.StoreTarget.writeTo(StoreTarget.java:45)
        at org.gradle.caching.http.internal.HttpBuildCacheService$1.writeTo(HttpBuildCacheService.java:151)
        at org.apache.http.impl.execchain.RequestEntityProxy.writeTo(RequestEntityProxy.java:121)
        at org.apache.http.impl.DefaultBHttpClientConnection.sendRequestEntity(DefaultBHttpClientConnection.java:156)
        at org.apache.http.impl.conn.CPoolProxy.sendRequestEntity(CPoolProxy.java:152)
        at org.apache.http.protocol.HttpRequestExecutor.doSendRequest(HttpRequestExecutor.java:238)
        at org.apache.http.protocol.HttpRequestExecutor.execute(HttpRequestExecutor.java:123)
        at org.apache.http.impl.execchain.MainClientExec.execute(MainClientExec.java:272)
        at org.apache.http.impl.execchain.ProtocolExec.execute(ProtocolExec.java:186)
        at org.apache.http.impl.execchain.RetryExec.execute(RetryExec.java:89)
        at org.apache.http.impl.client.InternalHttpClient.doExecute(InternalHttpClient.java:185)
        at org.apache.http.impl.client.CloseableHttpClient.execute(CloseableHttpClient.java:83)
        at org.gradle.internal.resource.transport.http.HttpClientHelper.performHttpRequest(HttpClientHelper.java:141)
        at org.gradle.internal.resource.transport.http.HttpClientHelper.performHttpRequest(HttpClientHelper.java:117)
        at org.gradle.caching.http.internal.HttpBuildCacheService.store(HttpBuildCacheService.java:159)
        ... 86 more
        Suppressed: javax.net.ssl.SSLException: Connection has been shutdown: javax.net.ssl.SSLException: java.net.SocketException: Broken pipe (Write failed)
                at sun.security.ssl.SSLSocketImpl.checkEOF(SSLSocketImpl.java:1554)
                at sun.security.ssl.SSLSocketImpl.checkWrite(SSLSocketImpl.java:1566)
                at sun.security.ssl.AppOutputStream.write(AppOutputStream.java:71)
                at org.apache.http.impl.io.SessionOutputBufferImpl.streamWrite(SessionOutputBufferImpl.java:124)
                at org.apache.http.impl.io.SessionOutputBufferImpl.flushBuffer(SessionOutputBufferImpl.java:136)
                at org.apache.http.impl.io.SessionOutputBufferImpl.flush(SessionOutputBufferImpl.java:144)
                at org.apache.http.impl.io.ContentLengthOutputStream.close(ContentLengthOutputStream.java:93)
                at com.google.common.io.Closer.close(Closer.java:214)
                at org.gradle.caching.internal.controller.service.StoreTarget.writeTo(StoreTarget.java:49)
                ... 100 more
        Caused by: javax.net.ssl.SSLException: java.net.SocketException: Broken pipe (Write failed)
                at sun.security.ssl.Alerts.getSSLException(Alerts.java:214)
                at sun.security.ssl.SSLSocketImpl.fatal(SSLSocketImpl.java:1967)
                at sun.security.ssl.SSLSocketImpl.fatal(SSLSocketImpl.java:1924)
                at sun.security.ssl.SSLSocketImpl.handleException(SSLSocketImpl.java:1888)
                at sun.security.ssl.SSLSocketImpl.handleException(SSLSocketImpl.java:1833)
                at sun.security.ssl.AppOutputStream.write(AppOutputStream.java:128)
                at org.apache.http.impl.io.SessionOutputBufferImpl.streamWrite(SessionOutputBufferImpl.java:124)
                at org.apache.http.impl.io.SessionOutputBufferImpl.flushBuffer(SessionOutputBufferImpl.java:136)
                at org.apache.http.impl.io.SessionOutputBufferImpl.write(SessionOutputBufferImpl.java:167)
                at org.apache.http.impl.io.ContentLengthOutputStream.write(ContentLengthOutputStream.java:113)
                at com.google.common.io.ByteStreams.copy(ByteStreams.java:112)
                at com.google.common.io.ByteSource.copyTo(ByteSource.java:246)
                at org.gradle.caching.internal.controller.service.StoreTarget.writeTo(StoreTarget.java:45)
                ... 100 more
        [CIRCULAR REFERENCE:java.net.SocketException: Broken pipe (Write failed)]
```

Unclear if there's a supported contract for this, but the Enterprise implementation does appear to support limiting the upload size - but I don't know how to make our implemention behave better than this.

### Expected Behavior
It probably makes sense for this configuration to be client side too but also wouldn't hurt to enable expect/continue support to allow a build cache to validate PUT before continuing:

https://hc.apache.org/httpclient-3.x/performance.html#Expect-continue_handshake

### Current Behavior
Limiting upload size appears to confuse Gradle.

### Context
Gradle 6.5

**Comment by lptr** [comment]
@ldaley please back me up here: was the reason for adding this feature on the server side to avoid filling up disk space too fast? In any case we never followed up with a client-side feature, though.

I wonder if there would be some way for the client to learn about the fact that the artifact they tried to upload was too big. There is HTTP 413, but I'm not sure how sending that on HTTP/1.x would work while the upload is in progress.

We could also implement HTTP 100 for large (what's large, though?) artifacts. It would not make sense to do it for all artifacts. I wonder if we can do this on the client side only, or if we need to make changes on the server-side, too.

We could also have some client-side configuration for this, or let the server-side provision the configuration somehow.

**Comment by ldaley** [comment]
> please back me up here: was the reason for adding this feature on the server side to avoid filling up disk space too fast?

Yes.

> I wonder if there would be some way for the client to learn about the fact that the artifact they tried to upload was too big. There is HTTP 413, but I'm not sure how sending that on HTTP/1.x would work while the upload is in progress.

Our server does respond with 413, as soon as it knows that it's going to be too big (which in practice is after receiving the headers as the client advertises the size). It will then immediately respond with a 413 and close the connection (as it's unsalvageable). The HTTP client we use doesn't deal with this situation well, at least with our usage of the synchronous API. It should be possible with the async API to detect this situation. 

> We could also implement HTTP 100 for large (what's large, though?) artifacts. It would not make sense to do it for all artifacts. I wonder if we can do this on the client side only, or if we need to make changes on the server-side, too.

Our server is already set up to handle expect-continue, so no changes needed there. We haven't tested with the Gradle client though.

> or let the server-side provision the configuration somehow.

For Gradle Enterprise, this is the direction I would like to go. We are moving towards a custom protocol for this connection that will support this. However, that doesn't help people not using Enterprise.

This issue tends to get complicated when we consider that dealing with large artifacts gets more nuanced when considering the cost/benefit of uploading them and how this should be controlled and communicated to users. When we have discussed this in the past, we have gotten stuck on this aspect.

**Comment by lptr** [comment]
Let's take a step back here. The idea about how to handle builds producing too-large artifacts is that in such a case one of two steps should be taken:

1. increase the accepted size if it makes sense to cache the artifact,
2. disable caching for the task, if it doesn't.

The current behavior on the client is to log the error, and then continue with a disable remote cache. This behavior does not support making the above choice, as it's pretty hard to notice the problem, and has unreasonable side-effects (why run the build with the remote cache disabled after this?)

It would be a better option to actually fail the build on HTTP 413, so the failure would be obvious, it would show up on the GE dashboards and build scans, and would not silently destroy build performance.

(We should do the same with HTTP 401 probably, i.e. when we cannot access the build cache server due to some configuration problem.)

**Comment by vlsi** [comment]
@lptr , there are cases when a task produces unexpectedly big cache entry.

For instance, I have a case where a single unit test consumes lots of megabytes in the cache.

https://github.com/pgjdbc/pgjdbc/blob/aa190cce2c27ac5f5d4fa33f5d156f8bf04cbe0c/pgjdbc/src/test/java/org/postgresql/core/UTF8EncodingTest.java#L29-L33

```java
  @Parameterized.Parameters(name = "string={1}, encoding={0}")
  public static Iterable<Object[]> data() {
    final StringBuilder reallyLongString = new StringBuilder(1024 * 1024);
    for (int i = 0; i < 185000; ++i) {
      reallyLongString.append(i);
    }
```

It turns out the test name goes into the cache entry, so the resulting cache entry becomes 35MiB.
I'm not sure it was the only test like that, but it does happen in practice.
Just in case: the test is using long string on purpose, so the only thing we can do here is to avoid putting the full string into the test description. Of course, we can do that, but it is not that obvious.

That is why I think it would be great to have

1) A configuration option to limit the maximum size of the cached entry.

    For instance, a global option + per-task override to protect users from unexpectedly large uploads/downloads during the build.

2) A configuration option to warn users if they produce tests with unexpectedly long test names.
I guess the reason the full test names appear in the build cache entry for the tests is that test task produces `output` with a representation of all the outcomes. That is why it is worth warning the users so they adjust the test names.

**Comment by lptr** [comment]
@vlsi I see how this was a hard-to-debug problem. I doubt it should be the responsibility of Gradle to handle this, though, and even if we want to do something here, it shouldn't be tied to the build cache.

One thing you can do to prevent these kinds of issues is to add a `doLast()` block to all of your `Test` tasks that checks the size of the output and fail the build if it finds it to be too large.

I think what we should do wrt Gradle's build cache size handling is to treat HTTP 413 properly by stopping the upload and continuing to use the remote cache for other tasks.

**Comment by vlsi** [comment]
>I doubt it should be the responsibility of Gradle to handle this

It is common practice to avoid doing something which would involve consuming too much resources.

>I think what we should do wrt Gradle's build cache size handling is to treat HTTP 413 properly

I guess it would be better to refrain from even trying to upload the entry. Of course, the handling for 413 would be nice anyway. What I mean it would be sad to start a multi-part upload the remote cache, upload 250GiB, and then receive 413.

**Comment by vlsi** [comment]
>One thing you can do to prevent these kinds of issues is to add a doLast() block to all of your Test tasks that checks the size of the output and fail the build if it finds it to be too large.

Do you mean only `test` tasks can have big cache entries? It would be nice to have a safety net for all the tasks, not just `test` ones.

**Comment by lptr** [comment]
We have a safety net which is the cache size limit. We could enforce this (=break the build if violated) or handle leniently (=handle HTTP 413 more gracefully). I think it makes sense to have a switch for this, so I think we agree here.

You brought up a use case for why a general limit is not a complete solution. I disagree that the surprising output of parametrized JUnit tests is a general-enough problem that should be handled at the level of the build cache, or Gradle in general. It is a very specific problem that could be handled in JUnit itself (like via the issue you reported there), or you could add something (like a `doLast()` check) in your build specifically for this.

**Comment by vlsi** [comment]
Eh. My JUnit sample was here to justify that tasks with unreasonable big outputs do exist in the wild. Of course, I don't expect Gradle to target a single case only, and I think adding a configuration option to limit the size of upload/download for the build cache would be a reasonable generic solution.

**Comment by lptr** [comment]
Alright. So we agree that an option to fail the build on HTTP 413 would be a good enough solution for all?

**Comment by vlsi** [comment]
No. I would rather skip the upload/download if it exceeds the limit.

Here's a relevant GitHub Actions caching code. They skip cache upload if they know that would overflow the limit. https://github.com/actions/toolkit/blob/905b2c7b0681b11056141a60055f1ba77358b7e9/packages/cache/src/cache.ts#L166-L175

```javascript
  const fileSizeLimit = 5 * 1024 * 1024 * 1024 // 5GB per repo limit
  const archiveFileSize = utils.getArchiveFileSizeIsBytes(archivePath)
  core.debug(`File Size: ${archiveFileSize}`)
  if (archiveFileSize > fileSizeLimit) {
    throw new Error(
      `Cache size of ~${Math.round(
        archiveFileSize / (1024 * 1024)
      )} MB (${archiveFileSize} B) is over the 5GB limit, not saving cache.`
    )
  }
```

**Comment by lptr** [comment]
I believe this is exactly what I'm proposing here. You'd have a _lenient_ mode where we'd react "nicely" to HTTP 413, and not disable caching for the rest of the build. We'd also have a _strict_ mode, where we'd break the build on HTTP 413. As @ldaley pointed out above, the HTTP 413 arrives as soon as the server has processed the headers, so there will be little data uploaded, and we waste no extra time on the client.

**Comment by vlsi** [comment]
lenient and strict would be great.

However, 413 can't always be the case.
For instance, AWS S3 has multipart upload API which initiates an upload, then client can send parts in parallel, then the upload is committed. The server has no idea on the total upload size at the initiation time.
See https://docs.aws.amazon.com/AmazonS3/latest/API/API_CreateMultipartUpload.html
So it might be helpful to have a client-side setting as well.

**Comment by lptr** [comment]
Are you referring to https://github.com/myniva/gradle-s3-build-cache? Build cache plugins like that can implement their own handling for cache size validation if they wish. The `BuildCacheEntryWriter` receives the size of the entry, so it should be easy to implement. They can add a property on their own implementation of `org.gradle.caching.configuration.BuildCache` to specify the max size if they want.

**Comment by vlsi** [comment]
> The BuildCacheEntryWriter receives the size of the entry, so it should be easy to implement

Ah. I've already implemented that in https://github.com/burrunan/gradle-s3-build-cache (`maximumCachedObjectLength`)
What I mean is there are multiple remote build cache implementations, and I guess it might be helpful to have a single Gradle-documented property to avoid 100500 slightly different options for essentially the same thing.

However, I do not think the discussion is worth time we spend :-/
I can live without cache limit at all.
I can live with lentient.
I can live with strict.
It would probably be more interesting to discuss https://github.com/gradle/gradle/issues/13982

I just think the global client-side option is easy to implement/document, and it might be helpful as well (e.g. users with low bandwidth could limit the amount of transfer the build does).

**Comment by vlsi** [comment]
Here's the relevant Slack discussion: https://gradle-community.slack.com/archives/CA83B1VLL/p1596193270151000

>Sinan Kozak Jul 31st at 2:01 PM
>Hello, in our android project one of the task has 65mb cache (mergeDebugNativeLibs). I wonder is it possible to exclude it >from remote cache but keep the local cache enabled? Avoidance saving is 14 seconds but 65mb is 1 of 3 or remote cache

**Comment by DanielThomas** [comment]
> However, 413 can't always be the case.

Gradle always sends a correct `content-length` header so it should be the case (we fallback to reading the body just in case - https://gist.github.com/DanielThomas/cdbbc12665534cce785c05a4c407bcce#file-gradle-cache-vcl-L21-L29).

In addition to configurable strictness around rejections, Expect/continue support so the client doesn't start sending the artifact until it has the thumbs up from the cache server would be a plus. We get that for free w/ our existing configuration.

**Comment by vlsi** [comment]
The issue is still relevant

**Comment by DanielThomas** [comment]
@vlsi there's opt-in Expect-Continue support now, solved as far as I'm concerned.

**Comment by ZacSweers** [comment]
Could you link docs on that?
